cdap-data-fabric/src/main/java/co/cask/cdap/data2/datafabric/dataset/service/executor/DatasetAdminService.java

```diff
@@ -104,20 +104,27 @@ public DatasetSpecification createOrUpdate(final Id.DatasetInstance datasetInsta
     }
     try (DatasetClassLoaderProvider classLoaderProvider =
            new DirectoryClassLoaderProvider(cConf, locationFactory)) {
+      final DatasetContext context = DatasetContext.from(datasetInstanceId.getNamespaceId());
+      UserGroupInformation ugi = impersonator.getUGI(datasetInstanceId.getNamespace().toEntityId());
 
-      return impersonator.doAs(datasetInstanceId.getNamespace().toEntityId(), new Callable<DatasetSpecification>() {
+      final DatasetType type = ImpersonationUtils.doAs(ugi, new Callable<DatasetType>() {
         @Override
-        public DatasetSpecification call() throws Exception {
+        public DatasetType call() throws Exception {
           DatasetType type = dsFramework.getDatasetType(typeMeta, null, classLoaderProvider);
           if (type == null) {
             throw new BadRequestException(
               String.format("Cannot instantiate dataset type using provided type meta: %s", typeMeta));
           }
-          DatasetSpecification spec = existing == null
-            ? type.configure(datasetInstanceId.getId(), props)
+          return type;
+        }
+      });
+
+      DatasetSpecification spec = ImpersonationUtils.doAs(ugi, new Callable<DatasetSpecification>() {
```

Comment:
the cost of doing this `doAs` is minimal. It doesn't involve any extra RPC or anything.